playground/backend/internal/cloud_bucket/cloud_storage.go

```diff
@@ -0,0 +1,96 @@
+// Licensed to the Apache Software Foundation (ASF) under one or more
+// contributor license agreements.  See the NOTICE file distributed with
+// this work for additional information regarding copyright ownership.
+// The ASF licenses this file to You under the Apache License, Version 2.0
+// (the "License"); you may not use this file except in compliance with
+// the License.  You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package cloud_bucket
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"strings"
+	"time"
+
+	"cloud.google.com/go/storage"
+	"google.golang.org/api/option"
+
+	pb "beam.apache.org/playground/backend/internal/api/v1"
+	"beam.apache.org/playground/backend/internal/logger"
+)
+
+const (
+	timeout = 5 * time.Second
+)
+
+type CloudStorage struct {
+	client *storage.Client
+}
+
+func NewCloudStorage(ctx context.Context) (*CloudStorage, error) {
+	client, err := storage.NewClient(ctx, option.WithoutAuthentication())
+	if err != nil {
+		return nil, err
+	}
+	return &CloudStorage{client: client}, nil
+}
+
+func (cs *CloudStorage) GetDatasets(ctx context.Context, bucketName string, datasets []*pb.Dataset) ([]*DatasetDTO, error) {
+	if cs.client == nil {
+		return nil, errors.New("storage client not set")
+	}
+
+	ctx, cancel := context.WithTimeout(ctx, timeout)
+	defer cancel()
+
+	bucket := cs.client.Bucket(bucketName)
+
+	var rc storage.Reader
```

Comment:
If it's an interface, it holds a reference to the value inside, no need to pass it explicitly below

Reply:
It's not an interface, it's a struct:

```
type Reader struct {
	Attrs              ReaderObjectAttrs
	seen, remain, size int64
	checkCRC           bool   // should we check the CRC?
	wantCRC            uint32 // the CRC32c value the server sent in the header
	gotCRC             uint32 // running crc

	reader io.ReadCloser
}
```